pkg/controller/atlasdatabaseuser/databaseuser.go

```diff
@@ -6,197 +6,313 @@ import (
 
-	if result := checkUserExpired(ctx.Context, ctx.Log, r.Client, project.ID(), dbUser); !result.IsOk() {
-		return result
+	r.dbUserService = dbuser.NewAtlasUsers(sdkClient.DatabaseUsersApi)
+	r.deploymentService = deployment.NewAtlasDeployments(sdkClient.ClustersApi, sdkClient.ServerlessInstancesApi, r.AtlasProvider.IsCloudGov())
+
+	return r.dbuLifeCycle(ctx, atlasDatabaseUser, atlasProject)
+}
+
+func (r *AtlasDatabaseUserReconciler) dbuLifeCycle(ctx *workflow.Context, atlasDatabaseUser *akov2.AtlasDatabaseUser, atlasProject *akov2.AtlasProject) ctrl.Result {
+	databaseUserInAtlas, err := r.dbUserService.Get(ctx.Context, atlasDatabaseUser.Spec.DatabaseName, atlasProject.ID(), atlasDatabaseUser.Spec.Username)
+	if err != nil && !errors.Is(err, dbuser.ErrorNotFound) {
+		return r.terminate(ctx, atlasDatabaseUser, api.DatabaseUserReadyType, workflow.Internal, true, err)
 	}
 
-	if err := validateScopes(ctx, ds, project.ID(), dbUser); err != nil {
-		return workflow.Terminate(workflow.DatabaseUserInvalidSpec, err.Error())
+	expired, err := isExpired(atlasDatabaseUser)
+	if err != nil {
+		return r.terminate(ctx, atlasDatabaseUser, api.DatabaseUserReadyType, workflow.DatabaseUserInvalidSpec, false, err)
 	}
+	if expired {
+		err = connectionsecret.RemoveStaleSecretsByUserName(ctx.Context, r.Client, atlasProject.ID(), atlasDatabaseUser.Spec.Username, *atlasDatabaseUser, r.Log)
+		if err != nil {
+			return r.terminate(ctx, atlasDatabaseUser, api.DatabaseUserReadyType, workflow.DatabaseUserConnectionSecretsNotDeleted, true, err)
+		}
 
-	if result := performUpdateInAtlas(ctx, r.Client, dus, project, &dbUser, apiUser); !result.IsOk() {
-		return result
+		return r.terminate(ctx, atlasDatabaseUser, api.DatabaseUserReadyType, workflow.DatabaseUserExpired, false, errors.New("an expired user cannot be managed"))
 	}
 
-	if result := checkDeploymentsHaveReachedGoalState(ctx, ds, project.ID(), dbUser); !result.IsOk() {
-		return result
+	scopesAreValid, err := r.areDeploymentScopesValid(ctx, atlasDatabaseUser, atlasProject)
+	if err != nil {
+		return r.terminate(ctx, atlasDatabaseUser, api.DatabaseUserReadyType, workflow.DatabaseUserInvalidSpec, false, err)
+	}
+	if !scopesAreValid {
+		return r.terminate(ctx, atlasDatabaseUser, api.DatabaseUserReadyType, workflow.DatabaseUserInvalidSpec, false, errors.New("\"scopes\" field refer to one or more deployments that don't exist"))
 	}
 
-	if result := connectionsecret.CreateOrUpdateConnectionSecrets(ctx, r.Client, ds, r.EventRecorder, project, dbUser); !result.IsOk() {
-		return result
+	dbUserExists := databaseUserInAtlas != nil
+	wasDeleted := !atlasDatabaseUser.DeletionTimestamp.IsZero()
+
+	switch {
+	case !dbUserExists && !wasDeleted:
+		return r.create(ctx, atlasDatabaseUser, atlasProject)
+	case dbUserExists && !wasDeleted:
+		return r.update(ctx, atlasDatabaseUser, atlasProject, databaseUserInAtlas)
+	case dbUserExists && wasDeleted:
+		return r.delete(ctx, atlasDatabaseUser, atlasProject)
+	default:
```

Comment:
!dbUserExists && wasDeleted -> not in Atlas and deleted in Kubernetes, no need to keep track of it anymore.